working-w-copilot/copilot-in-projects-refactoring.md

```diff
@@ -358,161 +313,153 @@ However, there are other scenarios that are worth testing. What happens if the r
 
+def test_update_book_missing_title(client, two_saved_books):
     # Act
-    response = client.put(f"/books/{book_id}", json=data)
+    response = client.put("/books/1", json={
+        "description": "Updated description"
+    })
+    response_body = response.get_json()
 
     # Assert
     assert response.status_code == 400
-    assert b"Title is required" in response.data
-
-def test_update_book_without_description(client, two_saved_books):
-    # Arrange
-    book_id = 1
-    data = {
-        "title": "Updated Book"
-    }
+    assert response_body == {"message": "Missing required field: title"}
 
+def test_update_book_missing_description(client, two_saved_books):
     # Act
-    response = client.put(f"/books/{book_id}", json=data)
+    response = client.put("/books/1", json={
+        "title": "Updated Ocean Book"
+    })
+    response_body = response.get_json()
 
     # Assert
     assert response.status_code == 400
-    assert b"Description is required" in response.data
+    assert response_body == {"message": "Missing required field: description"}
 ```
 
-Let's paste these tests into our test file and run them. The test `test_update_book_with_extra_keys` passes without changes, but the other two are failing. We need to take a deeper look at the tests for missing required data.
+Let's "Accept" and save these tests then run them. The test `test_update_book_with_extra_keys` should pass as-is, but the other two are failing. We need to take a deeper look at the tests for missing required data.
 
-The assertions for these tests assume that a 400 status code will be sent in case of an error, and also made an assumption about what the error message in the response would be. If we navigate to the `update_book` function in `routes.py`, we see that there actually isn't any error handling for this scenario. These tests fail because our application crashes with a `KeyError` when required data is missing from the request dictionary.
+The assertions for these tests assume that a 400 status code will be sent in case of an error, and also made an assumption about what the error message in the response would be. If we navigate to the `update_book` function in `book_routes.py`, we see that there actually isn't any error handling for this scenario. These tests fail because our application crashes with a `KeyError` when required data is missing from the request dictionary.
 
 While it's possible that Copilot could have written tests that expected a crash (which would look like a `500 Internal Server Error`), it's much more common that these scenarios would result in a `400 Bad Request`. So even though Copilot generated failing tests, it turns out that this helped us by highlighting error handling that would be useful to have. Rather than changing the tests to expect a crash, we're going to add error handling to `update_book` to make our app more robust, then update our test assertions to check for the correct message.
```

Comment:
Is this something that we should also plan to move earlier into the relevant lesson that has the PUT endpoint, similar to what we need to do to straighten out the POST situation?

Reply:
OK. I did mention this in the Asana ticket I linked to earlier.